src/v/cluster/tm_stm.cc

```diff
@@ -117,16 +117,77 @@ tm_stm::tm_stm(
       config::shard_local_cfg().tm_violation_recovery_policy.value())
   , _feature_table(feature_table) {}
 
-ss::future<std::optional<tm_transaction>>
-tm_stm::get_tx(const kafka::transactional_id& tx_id) {
+std::optional<tm_transaction>
+tm_stm::do_get_tx(const kafka::transactional_id& tx_id) {
     auto tx = _mem_txes.find(tx_id);
     if (tx != _mem_txes.end()) {
-        co_return tx->second;
+        return tx->second;
     }
     tx = _log_txes.find(tx_id);
     if (tx != _log_txes.end()) {
-        co_return tx->second;
+        return tx->second;
+    }
+    return std::nullopt;
+}
+
+ss::future<std::optional<tm_transaction>>
+tm_stm::get_tx(const kafka::transactional_id& tx_id) {
+    auto tx_opt = do_get_tx(tx_id);
+    if (!tx_opt) {
+        // not found.
+        co_return tx_opt;
+    }
+    auto tx = tx_opt.value();
+    if (
+      tx.status != tm_transaction::tx_status::ready
+      && tx.status != tm_transaction::tx_status::ongoing) {
+        co_return tx_opt;
+    }
+    // Check if transferring.
+    // We have 4 combinations here for
+    // transferring and etag/term match.
+    //
+    // transferring = tx->transferring
+    // term match = tx.etag == _insync_term
+    // +-------------------------+------+-------+
+    // | transferring/term match | True | False |
+    // +-------------------------+------+-------+
+    // | True                    |    1 |     2 |
+    // | False                   |    3 |     4 |
+    // +-------------------------+------+-------+
+
+    // case 1 - Invalid, we fail the invariant that the term is reset for
+    // transferring txn, abort.
+    //          This is likely a bug in the code.
+    // case 2 - Valid, txn is getting transferred from previous term.
+    // case 3 - Valid, the current term has already reset etag, so just return.
+    // case 4 - Invalid, txn can potentially contain dirty data from previous
+    // crashed leader, abort. We handle these 4 cases below.
+    auto term_match = tx.etag == _insync_term;
+    if (likely(!tx.transferring && term_match)) {
```

Comment:
> so we don't need to fail "invalid" cases.

So you are suggesting that we just return the tx and let the next client RPC deal with term mismatch?

Reply:
correct, we already handle the case when etag != _insync_term (by failing a tx) so we always may fallback to that existing behavior instead of failing with assert, see code snippet above